service/history/query.go

```diff
@@ -23,131 +23,125 @@
 package history
 
 import (
-	"sync"
+	"sync/atomic"
 
 	"github.com/pborman/uuid"
 
 	"github.com/uber/cadence/.gen/go/shared"
 )
 
 const (
-	queryStateBuffered queryState = iota
-	queryStateCompleted
-	queryStateUnblocked
+	queryTerminationTypeCompleted queryTerminationType = iota
+	queryTerminationTypeUnblocked
+	queryTerminationTypeFailed
 )
 
 var (
-	errQueryResultIsNil       = &shared.InternalServiceError{Message: "query result is nil"}
-	errQueryResultIsInvalid   = &shared.InternalServiceError{Message: "query result is invalid"}
-	errAlreadyInTerminalState = &shared.InternalServiceError{Message: "query already in terminal state"}
+	errTerminationStateInvalid = &shared.InternalServiceError{Message: "query termination state invalid"}
+	errAlreadyInTerminalState  = &shared.InternalServiceError{Message: "query already in terminal state"}
+	errQueryNotInTerminalState = &shared.InternalServiceError{Message: "query not in terminal state"}
 )
 
 type (
-	queryState int
+	queryTerminationType int
 
 	query interface {
-		getQueryInternalState() *queryInternalState
+		getQueryID() string
```

Comment:
This interface does not look very useful to me. You might want to just use the impl directly.

Reply:
I use an interface here because I want to make it clear to the reader that certain invariants of query are always return I cannot clearly indicate this if the rest of the system is working directly on references to query structs.